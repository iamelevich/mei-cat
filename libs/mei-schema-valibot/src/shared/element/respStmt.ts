import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile/attr/facsimile";
import { ModelNameLikeAgentSchema } from "../../namesdates/model/nameLike.agent";
import { AttrBiblSchema } from "../attr/bibl";
import { AttrCommonSchema } from "../attr/common";
import { ModelHeadLikeSchema } from "../model/headLike";
import { NameSchema } from "./name";
import { RespSchema } from "./resp";

/**
 * Base schema with attribute, to simplify types for RespStmtSchema
 */
const RespStmtBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrFacsimileSchema.entries,
});

/**
 * Transcription of text that names one or more individuals, groups, or in rare cases, mechanical processes, responsible for creation, realization, production, funding, or distribution of the intellectual or artistic content.
 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
 */
export const RespStmtSchema = v.lazy(() =>
	v.intersect([
		RespStmtBaseSchema,
		v.object({
			/**
			 * Reference to element name
			 * @see https://music-encoding.org/guidelines/v5/elements/name.html
			 */
			name: v.optional(v.union([v.array(NameSchema), NameSchema])),
			/**
			 * Reference to element resp
			 * @see https://music-encoding.org/guidelines/v5/elements/resp.html
			 */
			resp: v.optional(v.union([v.array(RespSchema), RespSchema])),
		}),
		ModelHeadLikeSchema,
		ModelNameLikeAgentSchema,
	]),
);

export type RespStmtData = v.InferOutput<typeof RespStmtSchema>;
