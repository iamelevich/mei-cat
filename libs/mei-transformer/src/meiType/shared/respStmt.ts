import { Type } from "typebox";
import { StandardTagSchema } from "../common";
import { AttrCommonSchema } from "./attr/common";
import { AttrLabelledSchema } from "./attr/labelled";
import { AttrLangSchema } from "./attr/lang";
import { AttrTypedSchema } from "./attr/typed";

/**
 * Transcription of text that names one or more individuals, groups, or in rare cases, mechanical processes,
 * responsible for creation, realization, production, funding, or distribution of the intellectual or artistic content.
 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
 */
export const RespStmtSchema = Type.Intersect([
	StandardTagSchema,
	AttrCommonSchema,
	AttrLangSchema,
	AttrLabelledSchema,
	AttrTypedSchema,
]);
