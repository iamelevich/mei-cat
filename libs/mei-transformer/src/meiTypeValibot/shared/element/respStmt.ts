import * as v from "valibot";
import { StandardTagSchema } from "../../common";
import { AttrFacsimileSchema } from "../../facsimile/attr/facsimile";
import { ModelNameLikeAgentSchema } from "../../namesdates/model/nameLike.agent";
import { AttrBiblSchema } from "../attr/bibl";
import { AttrCommonSchema } from "../attr/common";
import { NameSchema } from "../element/name";
import { RespSchema } from "../element/resp";
import { ModelHeadLikeSchema } from "../model/headLike";

/**
 * Base schema with attribute, to simplify types for RespStmtSchema
 */
const RespStmtBaseSchema = v.object({
	...StandardTagSchema.entries,
	...AttrBiblSchema.entries,
	...AttrCommonSchema.entries,
	...AttrFacsimileSchema.entries,
});

/**
 * Transcription of text that names one or more individuals, groups, or in rare cases, mechanical processes, responsible for creation, realization, production, funding, or distribution of the intellectual or artistic content.
 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
 */
export const RespStmtSchema = v.lazy(() =>
	v.intersect([
		RespStmtBaseSchema,
		v.object({
			/**
			 * Reference to element name
			 * @see https://music-encoding.org/guidelines/v5/elements/name.html
			 */
			name: v.optional(v.union([NameSchema, v.array(NameSchema)])),
			/**
			 * Reference to element resp
			 * @see https://music-encoding.org/guidelines/v5/elements/resp.html
			 */
			resp: v.optional(v.union([RespSchema, v.array(RespSchema)])),
		}),
		ModelHeadLikeSchema,
		ModelNameLikeAgentSchema,
	]),
);

export type RespStmtData = v.InferOutput<typeof RespStmtSchema>;
