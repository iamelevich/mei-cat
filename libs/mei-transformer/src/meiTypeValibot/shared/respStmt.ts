import * as v from "valibot";
import { StandardTagSchema } from "../common";
import { AttrCommonSchema } from "./attr/common";
import { AttrLangSchema } from "./attr/lang";

/**
 * Simplified schema for respStmt element.
 * @see https://music-encoding.org/guidelines/v5/elements/respStmt.html
 */
export const RespStmtSchema = v.object({
	...StandardTagSchema.entries,
	...AttrCommonSchema.entries,
	...AttrLangSchema.entries,
});

export type RespStmtData = v.InferOutput<typeof RespStmtSchema>;
