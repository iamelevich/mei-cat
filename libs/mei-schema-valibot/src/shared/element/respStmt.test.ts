import { describe, expect, it } from "bun:test";
import * as v from "valibot";
import { RespStmtSchema } from "./respStmt";

describe("RespStmtSchema", () => {
	it("should be defined", () => {
		expect(RespStmtSchema).toBeDefined();
	});

	it("should be a valid schema", () => {
		const parseResult = v.safeParse(RespStmtSchema, {
			name: {
				"#text": "ebp",
				"@xml:id": "resp_a27b49ae",
			},
			"@xml:id": "respStmt_d8b306d3",
		});

		expect(parseResult.success).toBe(true);
		expect(parseResult.output).toMatchSnapshot();
	});

	it("should be a valid shema with array of names", () => {
		const parseResult = v.safeParse(RespStmtSchema, {
			name: [
				{
					"#text": "ebp",
					"@xml:id": "resp_a27b49ae",
				},
				{
					"#text": "name 2",
					"@xml:id": "name_a27b49ae",
				},
			],
		});

		expect(parseResult.success).toBe(true);
		expect(parseResult.output).toMatchSnapshot();
	});
});
