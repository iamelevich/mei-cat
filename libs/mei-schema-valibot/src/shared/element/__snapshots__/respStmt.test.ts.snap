// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`RespStmtSchema should be a valid schema 1`] = `
{
  "@xml:id": "respStmt_d8b306d3",
  "name": {
    "#text": "ebp",
    "@xml:id": "resp_a27b49ae",
  },
}
`;

exports[`RespStmtSchema should be a valid shema with array of names 1`] = `
{
  "name": [
    {
      "#text": "ebp",
      "@xml:id": "resp_a27b49ae",
    },
    {
      "#text": "name 2",
      "@xml:id": "name_a27b49ae",
    },
  ],
}
`;
